src/stim/util_top/reference_sample_tree.inl

```diff
@@ -0,0 +1,161 @@
+                flush_recorded_into_result();
+                result.suffix_children.push_back(do_loop_with_tortoise_hare_folding(block, repeats));
+                start_size = sim.measurement_record.storage.size();
+            } else {
+                sim.do_gate(inst);
+            }
+        }
+    }
+
+    flush_recorded_into_result();
+    return result;
+}
+
+static uint64_t max_feedback_lookback_in_loop(const Circuit &loop) {
+    uint64_t furthest_lookback = 0;
+    for (const auto &inst : loop.operations) {
+        if (inst.gate_type == GateType::REPEAT) {
+            furthest_lookback = std::max(furthest_lookback, max_feedback_lookback_in_loop(inst.repeat_block_body(loop)));
+        } else {
+            auto f = GATE_DATA[inst.gate_type].flags;
+            if ((f & GateFlags::GATE_CAN_TARGET_BITS) && (f & GateFlags::GATE_TARGETS_PAIRS)) {
+                // Feedback-capable operation. Check for any measurement record targets.
+                for (auto t : inst.targets) {
+                    if (t.is_measurement_record_target()) {
+                        furthest_lookback = std::max(furthest_lookback, (uint64_t)-t.rec_offset());
+                    }
+                }
+            }
+        }
+    }
+    return furthest_lookback;
+}
+
+template <size_t W>
+ReferenceSampleTree CompressedReferenceSampleHelper<W>::do_loop_with_tortoise_hare_folding(const Circuit &loop, uint64_t reps) {
+    if (reps < 10) {
+        return do_loop_with_no_folding(loop, reps);
+    }
+
+    ReferenceSampleTree result;
+    result.repetitions = 1;
+
+    CompressedReferenceSampleHelper<W> tortoise(sim);
+    CompressedReferenceSampleHelper<W> hare(std::move(sim));
+    uint64_t max_feedback_lookback = max_feedback_lookback_in_loop(loop);
+    uint64_t tortoise_steps = 0;
+    uint64_t hare_steps = 0;
+    while (hare_steps < reps) {
+        hare_steps++;
+        result.suffix_children.push_back(hare.do_loop_with_no_folding(loop, 1));
+        assert(result.suffix_children.size() == hare_steps);
+
+        if (tortoise.in_same_recent_state_as(hare, max_feedback_lookback, hare_steps < 10)) {
+            break;
+        }
+
+        // Tortoise advances half as quickly.
+        if (hare_steps & 1) {
+            tortoise_steps++;
+            tortoise.do_loop_with_no_folding(loop, 1);
+        }
+    }
+
+    if (hare_steps == reps) {
+        // No loop found.
+        sim = std::move(hare.sim);
+        return result;
+    }
+
+    // Advance until the remaining iterations are a multiple of the period.
+    assert(result.suffix_children.size() == hare_steps);
+    uint64_t period = hare_steps - tortoise_steps;
+    size_t period_steps_left = (reps - hare_steps) / period;
+    while ((reps - hare_steps) % period) {
+        result.suffix_children.push_back(hare.do_loop_with_no_folding(loop, 1));
+        hare_steps += 1;
+    }
+    assert(hare_steps + period_steps_left * period == reps);
```

Comment:
Nit: Parenthesis would sure make this easier to reason about quickly.

Reply:
What do you mean? Parens where? Like, around the multiplication? IMO that's overkill.

My subjective experience of looking at this equation is that the grouping is extremely clear. More objectively: `+`, `*`, and `==` are extremely common operations that people have practice PEDMAS-ing on from elementary school through university.

Reply:
But precedence from left to right isn't part of PEMDAS (I learned PEMDAS in elementry school, not PEDMAS. No difference, but hey).

This idea is that cognitive overhead should be minimal. If someone is reading code dealing with some encoding of bits into a tree structure, being distracted by operator precedence is pretty silly.

In this case, yes, `*` takes precedence. But, it being on the right does add a moment to double check and make sure that works out in your head. Just glossing over these sorts of things (not being critical and actually thinking "wait, is that the right order?") is how bugs happen. Adding parenthesis removes the cognitive overhead, removes punishment for detailed readers, and removes the room for bugs.

> `(a + (b*c)) == d`

While, yes, this is exactly what I would write, I would at least encourage `a + (b*c) == d` to address this specific issue of cognitive overhead.

I have seen countless times that someone firmly believes in the benefits of terseness, that parenthesis just aren't necessary, thinks that they know operator precedence, get it wrong, and create a hard to find bug. I always argue that verbosity is not to be feared and it can often be easier to read and less bug prone.

Fun thing to share: https://gcc.gnu.org/git/?p=gcc.git&a=commit;h=23462d4d8fb5b5e40d6a62dd285301228f13a37b
> Properly handle NAND case by calculating ~(t1 & val) instead of (~t1 & val).

Yes, the impl details are different/more complex than a simple case of not/using parenthesis, but the point is made. And now this a permanent scar on GCC via the very annoying `-Wsync-nand`.

Reply:
Weird how these subjective experiences radically differ. Anyways imma leave it as is.